subsys/usb/class/netusb/function_rndis.c

```diff
@@ -0,0 +1,1000 @@
+			 * hoping to get buffers for later ones
+			 */
+			rndis.skip_bytes = rndis.in_pkt_len - len;
+
+			SYS_LOG_ERR("Not enough net buffers, len %u, skip %u",
+				    rndis.in_pkt_len, rndis.skip_bytes);
+
+			net_pkt_unref(pkt);
+			rndis.rx_no_buf++;
+			return;
+		}
+
+		net_pkt_frag_insert(pkt, buf);
+
+		rndis.in_pkt = pkt;
+
+		/* Append data only, skipping RNDIS header */
+		hdr_offset = sizeof(struct rndis_payload_packet);
+	}
+
+	ret = net_pkt_append_all(rndis.in_pkt, len - hdr_offset,
+				 buffer + hdr_offset, K_FOREVER);
+	if (ret < 0) {
+		SYS_LOG_ERR("Error appending data to pkt: %p", rndis.in_pkt);
+		rndis_clean();
+		rndis.rx_err++;
+		return;
+	}
+
+	SYS_LOG_DBG("To asemble %d bytes, reading %u bytes",
+		    rndis.in_pkt_len, len);
+
+	rndis.in_pkt_len -= len;
+	if (!rndis.in_pkt_len) {
+		SYS_LOG_DBG("Assembled full RNDIS packet");
+
+		net_hexdump_frags(">", rndis.in_pkt);
+
+		/* Queue data to iface */
+		if (net_recv_data(rndis.iface, rndis.in_pkt) < 0) {
+			net_pkt_unref(rndis.in_pkt);
+		}
+
+		/* Start over for new packets */
+		rndis.in_pkt = NULL;
+	} else if (rndis.in_pkt_len < 0) {
+		SYS_LOG_ERR("Error assembling packet, drop and start over");
+		rndis_clean();
+	}
+}
+
+static void rndis_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+{
+#ifdef VERBOSE_DEBUG
+	SYS_LOG_DBG("EP 0x%x status %d", ep, ep_status);
+#endif
+}
+
+static void rndis_bulk_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+{
+#ifdef VERBOSE_DEBUG
+	SYS_LOG_DBG("EP 0x%x status %d", ep, ep_status);
+#endif
+}
+
+static void rndis_notify(struct k_work *work)
+{
+	u32_t buf[2];
+
+	SYS_LOG_DBG("");
+
+	buf[0] = sys_cpu_to_le32(0x01);
+	buf[1] = sys_cpu_to_le32(0x00);
+
+	try_write(CONFIG_RNDIS_INT_EP_ADDR, (u8_t *)buf, sizeof(buf));
+
+	/* Decrement notify_count here */
+	if (atomic_dec(&rndis.notify_count) != 1) {
+		SYS_LOG_WRN("Queue next notification, count %u",
+			    atomic_get(&rndis.notify_count));
+
+		k_delayed_work_submit(&notify_work, K_NO_WAIT);
+	}
+}
+
+static void rndis_queue_rsp(struct net_buf *rsp)
+{
+
+	if (!k_fifo_is_empty(&rndis_tx_queue)) {
+#if 0
+		struct net_buf *buf;
+
+		while ((buf = net_buf_get(&rndis_tx_queue, K_NO_WAIT))) {
+			SYS_LOG_ERR("Drop buffer %p", buf);
+			net_buf_unref(buf);
+		}
+#endif
+		SYS_LOG_WRN("Transmit response queue is not empty");
+	}
+
+	net_buf_put(&rndis_tx_queue, rsp);
+}
+
+/* Notify host about available data */
+static void rndis_notify_rsp(void)
+{
+	int ret;
+
+	SYS_LOG_DBG("");
+
+	/* Keep track of number of notifies */
+	if (atomic_inc(&rndis.notify_count) != 0) {
+		SYS_LOG_WRN("Unhandled notify: count %u",
+			    atomic_get(&rndis.notify_count));
+
+		return;
+	}
+
+	/*
+	 * TODO: consider delay
+	 * k_delayed_work_submit(&notify_work, K_MSEC(1));
+	 */
+	ret = k_delayed_work_submit(&notify_work, K_NO_WAIT);
+	if (ret) {
+		SYS_LOG_ERR("Error submittinf delaying queue: %d", ret);
+	}
+}
+
+struct tlv {
+	u32_t type;
+	u32_t len;
+	u8_t data[0];
```

Comment:
Do we need to support *very* old GCC compilers? Might be better to just declare `data` as `u8_t data[]`, that is standard C99.

Reply:
ok